---
title: "Read alignments and coverage plots on a selected genomic region using Gviz"
author: "Marc Galland"
date: "`r Sys.Date()`"
output:
    html_document:
        number_sections: yes
        toc: yes
        toc_depth: 2  
---

# Setup
```{r setup, warning=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo=TRUE,message=FALSE,warning=FALSE,cache=FALSE,fig.align="center")
```

## Load libraries and set current working directory
The `IRanges`, `GenomicRanges`, `GenomicAlignments` and `GenomicFeatures` are important packages that define classes to store genomic information
(genomic coordinates, annotations, alignments...)
The `Gviz`package allows plotting of these genomic information and NGS read alignments. 
biomaRt allows to retrieve a lot of information from Ensembl (gene annotations, positions on chromosomes, GO, KEGG pathways...)
```{r Load librairies, echo=TRUE,results='hide',message=FALSE,warning=FALSE}
library(reshape2)
library(dplyr)
library(yaml)
library(Rsamtools)
library(IRanges)
library(GenomicRanges)
library(GenomicAlignments)
library(GenomicFeatures)
library(Gviz)
library(biomaRt)
library(parallel)
```

## Load configuration file
```{r}
config = yaml::yaml.load_file("config.yaml")
```

## Select region of interest
__Important__, also specify the DNA strand of interest too!
````{r region of interest}
# select only region of interest
selectedGenome = config$region$genome
selectedChrom = config$region$chr
selectedStart = config$region$start
selectedEnd = config$region$end
selectedStrand = config$region$strand
```

# Create the different tracks for ploting

## Create genome axis
```{r genome axis}
axisTrack = GenomeAxisTrack(name = "coordinates",range = IRanges(start = selectedStart,end = selectedEnd))
```

## Build Gene Region Track 
Here I construct the Gene Region Track manually. Can also be done using a TxDb object (see at the end of this Rmarkdown document)
```{r gene region track}
# import GTF fata 
gtf = rtracklayer::import.gff(config$annotations$gtf)

# range of interest
#interval2keep = GRanges(seqnames = selectedChrom,IRanges(selectedStart,selectedEnd),strand = config$region$strand)

# subset GRanges gtf object using selected start and end positions
df =  gtf[seqnames(gtf)==selectedChrom]
df = df[start(df) >= selectedStart]
df = df[end(df) <= selectedEnd]

# keep only protein coding features
#df = df[elementMetadata(df)$transcript_biotype == "protein_coding"]
  
# add gene symbol for plotting
#genes = cbind.data.frame(gene_id = config$genes$solycids,symbol = config$genes$symbol)
#df = dplyr::left_join(df,genes,by="gene_id")

# creates the GeneRegionTrack
txTrack = GeneRegionTrack(
  df,
  feature = df$type,
  symbol = df$transcript_id,
  name = "Gene Annotation",
  background.title="cadetblue3",
  shape=c("smallArrow"),
  strand=df$strand,
  stacking = "dense",
  showId = TRUE,
  options(ucscChromosomeNames=FALSE),
  genome=selectedGenome
  )
```

## Build Data Tracks that contain coverage values for range of interest
```{r}
# # directory for result files
# bamdir = config$bamdir
# 
# # List bedgraph files
# bedgraphFiles = sapply(list.files(config$bamdir,pattern = ".bedgraph$"),FUN = function(x) {file.path(bamdir,x)})
# bedgraphsNames = sapply(names(bedgraphFiles),FUN = function(x){strsplit(x,split = "\\.")[[1]][[1]]})
# bedgraphFiles = lapply(bedgraphFiles,function(x){read.delim(x,header = F,col.names = c("chromosome","start","end","cov"))})
# 
# # build Data Tracks
# BedTracks = lapply(bedgraphFiles,function(x){DataTrack(x,ctype="l",genome = selectedGenome)}) 
```

## Build Alignment Tracks from BAM files
```{r build alignment tracks}
# list all bam files of interest
bamdir = config$bamdir
bamFiles = sapply(list.files(config$bamdir,pattern = ".bam$"),FUN = function(x) {file.path(bamdir,x)})

# read BAM files and get their names
bamFiles = BamFileList(bamFiles)
bamFilesNames = lapply(bamFiles,
                       FUN = function(x){
                         strsplit(basename(path(x)),split = "\\.")[[1]][[1]]
                         })

# read alignments and store them in GRanges objects
readGAs = lapply(bamFiles,FUN = function(x) {readGAlignmentPairs(x,use.names = T)})

# load GAalignmentsPairs and convert to GRanges
readGRs = lapply(readGAs,FUN = function(x){granges(x,use.names = T,use.mcols = T)})

# creates AlignmentTracks
bamAlnTracks = lapply(readGRs,function(x){AlignedReadTrack(x,start = selectedStart,end = selectedEnd,chromosome = selectedChrom,genome = selectedGenome)})
names(bamAlnTracks) = bamFilesNames
```

# Merge all tracks and plot
```{r merge tracks}
# build track list and plot
trackList = append(x = list(axisTrack,txTrack),values = bamAlnTracks)
plotTracks(trackList,from = selectedStart,to = selectedEnd,transcriptAnnotation=c("symbol"),type="coverage")

# save to file
png(filename = "plot.png",width = 10,height = 6,res = 400,pointsize = 8)
plotTracks(trackList,from = selectedStart,to = selectedEnd,transcriptAnnotation=c("symbol"))
dev.off()
```


## Build custom ideogram track
```{r build custom ideogram}
ideoTrack <- IdeogramTrack(genome = "hg38",chromosome = "chr22")

# set your own
customIdeoTrack <- ideoTrack

# RANGE
START=seq(1,100,10)
END=seq(10,100,10)
RANGES <- GRanges(seqnames = c(letters[1:10]),IRanges(start =START ,end =END))

# arbitrary Giemsa staining
GS <- rep(c("mireille","marcelle","jacqueline","franÃ§ois","francis","gerard","robert","eric","nicolas","maurice"),times = 2)
bt <- data.frame(chrom=selectedChrom,chromStart= selectedStart,chromEnd=seq(10,100,10),name=letters[1:10],gieStain=GS)

# chromosome name
CHR <- "own1"

# Replace with your own data
my_ideoTrack@range <- RANGES
my_ideoTrack@chromosome <- CHR
my_ideoTrack@name <- CHR
my_ideoTrack@bandTable <- bt

# set UCSC chromosome Names False
options(ucscChromosomeNames=FALSE)
gx <- GenomeAxisTrack()
# Plot
plotTracks(list(my_ideoTrack,gx),from = 1,to = 15)
```

# Session info
```{r}
sessionInfo()
```

# not used but kept for documentation
```{r backup}
# ideoTrack <- IdeogramTrack(genome = "hg38",chromosome = "chr22")
# 
# # set your own
# my_ideoTrack <- ideoTrack
# 
# # RANGE
# START=seq(1,100,10)
# END=seq(10,100,10)
# RANGES <- GRanges(seqnames = c(letters[1:10]),IRanges(start =START ,end =END))
# 
# # arbitrary Giemsa staining
# GS <- c("gneg", "gpos100",  "gpos25","acen",  "gpos50",  "gpos75",    "gvar",   "stalk" ,NA,NA)
# bt <- data.frame(chrom="own1",chromStart= seq(1,100,10),chromEnd=seq(10,100,10),name=letters[1:10],gieStain=GS)
# 
# # chromosome name
# CHR <- "own1"
# 
# # Replace with your own data
# my_ideoTrack@range <- RANGES
# my_ideoTrack@chromosome <- CHR
# my_ideoTrack@name <- CHR
# my_ideoTrack@bandTable <- bt
# 
# # set UCSC chromosome Names False
# options(ucscChromosomeNames=FALSE)
# gx <- GenomeAxisTrack()
# # Plot
# plotTracks(list(my_ideoTrack,gx),from = 1,to = 15)
```

## Create a TxDb object
The TxDB object (GenomicFeatures package) is used to store genome annotation information. For instance, we can retrieve exons from transcripts, 3'UTR etc.
```{r}
# create a TxDb object (GenomicFeatures package) to store genome annotation information 
#txdb = makeTxDbFromGFF(file = config$annotations$gff,format = "gff3",dataSource = config$annotations$gffSource,organism = "Solanum lycopersicum")
````

## Use biomaRt to retrieve information
```{r biomart}
# # list available datasets from ensembl plants
# head(listDatasets(useMart("plants_mart",host="plants.ensembl.org")),20)
# 
# # list attributes available in a selected mart (plants_mart)
# head(listAttributes(useDataset(
#   dataset = "slycopersicum_eg_gene",
#   mart = useMart("plants_mart",host = "plants.ensembl.org")))
#   ,10)
# 
# # list filters that can be applied to the dataset
# head(listFilters(useDataset(dataset = "slycopersicum_eg_gene", 
#                             mart    = useMart("plants_mart",
#                                               host = "plants.ensembl.org"))), 100)
# 
# # Data retrieval
# plants = useMart("plants_mart",host = "plants.ensembl.org")
# tomatoMart = useDataset("slycopersicum_eg_gene",mart = plants) 
# 
# # get all genes + their annoations
# listOfDesiredAttributes = c("ensembl_gene_id",
#                             "ensembl_transcript_id",
#                             "chromosome_name",
#                             "start_position",
#                             "end_position",
#                             "strand",
#                             "description",
#                             "kegg_id",
#                             "uniprot_swissprot_accession",
#                             "go_accession",
#                             "go_name_1006",
#                             "interpro_id",
#                             "interpro_short_description")
# tomatogenes = getBM(attributes = listOfDesiredAttributes,mart=tomatoMart)
```